File: FdP-B/02/code/prodotto.cpp
#include <iostream>
#include <string>
using namespace std;

class Prodotto {
private:
    int codice;
    string nome;
    double prezzo;
    int quantita;
    bool disponibilita;

public:
    Prodotto() : Prodotto(0, "", 0, 0) { }
    Prodotto(int codice, string nome) : Prodotto(codice, nome, 0, 0){ }
    Prodotto(int codice, string nome, double prezzo, int quantita){
        this->codice = codice;
        this->nome = nome;
        this->prezzo = prezzo;
        this->quantita = quantita;
        this->disponibilita = quantita > 0;
    }

    int getCodice() const { return this->codice; }
    string getNome() const { return this->nome; }
    double getPrezzo() const { return this->prezzo; }
    int getQuantita() const { return this->quantita; }
    bool isDisponibile() const { return this->disponibilita; }

    void setCodice(const int& codice) {
        if (codice < 0)
            throw string("Errore: il codice deve essere positivo");
        
        this->codice = codice;
    }

    void setNome(const string& nome) { this->nome = nome; }

    void setPrezzo(const double& prezzo) {
        if (prezzo < 0) 
            throw string("Errore: il prezzo deve essere positivo");
        this->prezzo = prezzo;
    }

    void setQuantita(const int& quantita) {
        if (quantita < 0)
            throw string("Errore: la quantita deve essere positiva");
        
        this->quantita = quantita;
        this->disponibilita = quantita > 0;
    }

    void aggiungiQuantita() { aggiungiQuantita(1); }

    void aggiungiQuantita(int quantita) {
        if (quantita <= 0)
            throw string("Errore: la quantita deve essere positiva");
        
        this->quantita += quantita;
        this->disponibilita = this->quantita > 0;
    }

    void rimuoviQuantita() { rimuoviQuantita(1); }

    void rimuoviQuantita(int quantita) {
        if (this->quantita == 0) 
            throw string("Errore: la quantita e' gia' zero");
        
        if (quantita > this->quantita) 
            throw string("Errore: la quantita non puo' essere negativa");
        
        this->quantita -= quantita;
        this->disponibilita = this->quantita > 0;
    }

    Prodotto operator+(const Prodotto& other) const {
        if(this->codice == other.getCodice()){
            Prodotto p(this->codice, 
                        this->nome, 
                        this->prezzo, 
                        this->quantita + other.getQuantita());
            return p;
        }
        throw string("Errore: i due oggetti hanno codice diverso, impossibile sommarli");
    }

    bool operator==(const Prodotto& other) const {
        return this->codice == other.getCodice();
    }
    bool operator!=(const Prodotto& other) const {
        return this->codice != other.getCodice();
    }

    // Operatore di incremento postfisso
    Prodotto operator++(int valore) {
        Prodotto result(this->codice, this->nome, this->prezzo, this->quantita);
        this->aggiungiQuantita();
        return result;
    }

    // Operatore di incremento infisso
    Prodotto operator++() {
        Prodotto result(this->codice, this->nome, this->prezzo, this->quantita);
        result.aggiungiQuantita();
        this->aggiungiQuantita();
        return result;
    }

    void print() const {
        cout << "Nome: " << this->nome << " [" 
            << this->codice << "], € " << this->prezzo << ", quantità: " << this->quantita << endl;
    }
};

ostream& operator<<(ostream& os, const Prodotto& p) {
    os << p.getNome() << " [id=" 
        << p.getCodice() << "], €" 
        << p.getPrezzo() << ", quantità: " 
        << p.getQuantita() << ", disponibile: "
        << (p.isDisponibile() ? "si" : "no") << endl;
    return os;
}

void testProdotto() {
    cout << endl << " ==== Test classe Prodotto ==== " << endl;

    cout << endl << " ==== Costruttori ==== " << endl;
    Prodotto p1;
    p1.print();

    Prodotto p2(1, "Tastiera");
    p2.print();

    Prodotto p3(2, "Mouse", 20, 5);
    p3.print();

    cout << endl << " ==== Test setter ==== " << endl;
    p3.setQuantita(10);
    p3.print();

    cout << endl << " ==== Test aggiungiQuantita ==== " << endl;
    p2.aggiungiQuantita(20);
    p2.print();
    p2.aggiungiQuantita();
    p2.print();

    cout << endl << " ==== Test rimuoviQuantita ==== " << endl;
    p2.rimuoviQuantita(8);
    p2.print();

    cout << endl << " ==== Test rimuoviQuantita con errore ==== " << endl;
    try {
        p2.rimuoviQuantita(100);
        p2.print();
    } catch (const string& e) {
        cerr << e << endl;
    }

    cout << endl << " ==== Test overload operatore di output (<<) ==== " << endl;
    cout << p2;

    cout << endl << " ==== Test con errore overload operatore di somma (+) ==== " << endl;
    try {
        Prodotto p4 = p1 + p3;
        cout << p4 << endl;
    } catch (const string& e) {
        cerr << e << endl;
    }

    cout << endl << " ==== Test overload operatore di somma (+) ==== " << endl;
    Prodotto p5(2, "Mouse da gaming", 20, 60);
    Prodotto p6 = p5 + p3;
    cout << p6;

    cout << endl << " ==== Test overload operatore di confronto (==) ==== " << endl;
    cout << "True: " << (p3 == p5) << endl;
    cout << "False: " << (p1 == p2) << endl;

    cout << endl << " ==== Test overload operatore di confronto (!=) ==== " << endl;
    cout << "False: " << (p3 != p5) << endl;
    cout << "True: " << (p1 != p2) << endl;
    
    cout << endl << " ==== Test overload operatore di incremento postfisso (++) ==== " << endl;
    cout << "p3 pre incremento: " << p3;
    Prodotto p4 = p3++;
    cout << "p4: " << p4;
    cout << "p3 post incremento: " << p3;

    cout << endl << " ==== Test overload operatore di incremento infisso (++) ==== " << endl;
    cout << "p3 pre incremento: " << p3;
    p4 = ++p3;
    cout << "p4: " << p4;
    cout << "p3 post incremento: " << p3;

    cout << endl << " ==== Test incremento ==== " << endl;
    cout << "p3 pre incremento: " << p3;
    p4++;
    cout << "p3 post incremento: " << p3;
    cout << "p4 post incremento: " << p4;

    cout << endl << " ==== Fine Test classe Prodotto ==== " << endl;
}

int main() {
    testProdotto();
    return 0;
}

File: FdP-B/03/code/Prodotto.java
package it.unipr.fdpb.lecture03;

public class Prodotto {
    private int codice;
    private String nome;
    private double prezzo;
    private int quantita;
    private Boolean disponibilita;

    public Prodotto() {
        this(0, "", 0, 0);
    }

    Prodotto(int codice, String nome) {
        this(codice, nome, 0, 0);
    }

    public Prodotto(int codice, String nome, double prezzo, int quantita) {
        this.codice = codice;
        this.nome = nome;
        this.prezzo = prezzo;
        this.quantita = quantita;
        this.disponibilita = quantita > 0;
    }

    public int getCodice() {
        return this.codice;
    }
    public String getNome() {
        return this.nome;
    }
    public double getPrezzo() {
        return this.prezzo;
    }
    public int getQuantita() {
        return this.quantita;
    }
    public Boolean isDisponibile(){
        return this.disponibilita;
    }

    public void setCodice(int codice) {
        if (codice < 0)
            throw new RuntimeException("Codice invalido, deve essere positivo");
        this.codice = codice;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public void setPrezzo(double prezzo) {
        if (prezzo < 0)
            throw new RuntimeException("Prezzo invalido, deve essere positivo");
        this.prezzo = prezzo;
    }

    public void setQuantita(int quantita) {
        if (quantita < 0)
            throw new RuntimeException("Quantita invalida, deve essere positiva");
        this.quantita = quantita;
        this.disponibilita = quantita > 0;
    }

    public void aggiungiQuantita(int quantita) {
        if (quantita < 0)
            throw new RuntimeException("Quantita invalida, deve essere positiva");
        this.quantita += quantita;
        this.disponibilita = quantita > 0;
    }

    public void aggiungiQuantita() {
        aggiungiQuantita(1);
    }

    public void rimuoviQuantita(int quantita){
        if (quantita < 1)
            throw new RuntimeException("Impossibile rimuovere perche la quantita' e' gia' zero");

        if(quantita > this.quantita)
            throw new RuntimeException("Quantita invalida, deve essere minore di quella esistente");

        this.quantita -= quantita;
        this.disponibilita = this.quantita > 0;
    }

    public void rimuoviQuantita() {
        rimuoviQuantita(1);
    }

    public Prodotto somma(Prodotto other) {
        if (this.codice != other.getCodice())
            throw new RuntimeException("Impossibile sommare, codici diversi");

        return new Prodotto(
                this.codice,
                this.nome,
                this.prezzo,
                this.quantita + other.getQuantita());
    }

    public boolean equals(Prodotto other) {
        return this.codice == other.getCodice();
    }

    public void print(){
        System.out.println(nome + " [" + codice + "], €" + prezzo + " [quantita': " + quantita + "]");
    }

    @Override
    public String toString() {
        return nome + " [" + codice + "], €" + prezzo + " [quantita': " + quantita + "]";
    }

    public static void main(String[] args) {
        Prodotto p1 = new Prodotto();
        System.out.println(p1);

        p1.setNome("Prodotto predefinito");
        p1.print();

        Prodotto p2 = new Prodotto(1, "Tastiera");
        p2.print();

        Prodotto p3 = new Prodotto(2, "Mouse", 20.0, 5);
        p3.print();

        p2.setQuantita(10);
        p2.setPrezzo(5);
        p2.print();

        p3.aggiungiQuantita();
        p3.aggiungiQuantita(7);
        p3.print();

        p2.rimuoviQuantita();
        p2.rimuoviQuantita(4);
        p2.print();

        try {
            p1.rimuoviQuantita(100);
        } catch (RuntimeException e) {
            System.err.println(e.getMessage());
        }

        Prodotto p4 = new Prodotto(1, "Tastiera da gaming", 7, 100);
        p4.print();

        try {
            Prodotto p5 = p1.somma(p4);
        } catch (RuntimeException e) {
            System.err.println(e.getMessage());
        }

        Prodotto p5 = p2.somma(p4);
        p5.print();

        if (p2.equals(p4))
            System.out.println("Equal");

        if (!p2.equals(p3))
            System.out.println("Not equal");
    }
}
